File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/data/daos/impl/PlayerDao.java
package com.dkellycollins.triggerfinger.data.daos.impl;

import com.dkellycollins.triggerfinger.data.daos.IPlayerDao;
import com.dkellycollins.triggerfinger.data.entity.IPlayer;
import com.dkellycollins.triggerfinger.data.entity.impl.Player;

import java.util.HashMap;

public class PlayerDao extends BaseEntityDao<Player> implements IPlayerDao {

    private Player _playerOne;

    public PlayerDao() {
        super("PlayerDao");
    }

    @Override
    public int create(int collidableId, int invincibleTimerId, int weaponId, int score, byte health) {
        int id = getNextId();
        Player player = new Player(id, collidableId, invincibleTimerId, weaponId, score, health);

        _store.put(id, player);
        if(_playerOne == null) {
            _playerOne = player;
        }

        return id;
    }

    @Override
    public Iterable<IPlayer> retrieve() {
        return (Iterable<IPlayer>) (Iterable<?>) _store.values();
    }

    @Override
    public IPlayer retrieve(int id) {
        if(!_store.containsKey(id)) {
            throw new RuntimeException(String.format("Key {0} not found", id));
        }

        return _store.get(id);
    }

    @Override
    public IPlayer playerOne() {
        return _playerOne;
    }

    @Override
    public void update(int id, int weaponId, int score, byte health) {
        if(!_store.containsKey(id)) {
            throw new RuntimeException(String.format("Key {0} not found", id));
        }

        Player player = _store.get(id);

        player.setWeaponId(weaponId);
        player.setScore(score);
        player.setHealth(health);
    }

    @Override
    public void delete(int id) {
        if(!_store.containsKey(id)) {
            throw new RuntimeException(String.format("Key {0} not found", id));
        }

        _store.remove(id);
        if(_playerOne.getId() == id) {
            _playerOne = null;
        }
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/data/daos/impl/PlayerDao.kt
package com.dkellycollins.triggerfinger.data.daos.impl

import com.dkellycollins.triggerfinger.data.daos.IPlayerDao
import com.dkellycollins.triggerfinger.data.entity.IPlayer
import com.dkellycollins.triggerfinger.data.entity.impl.Player

import java.util.HashMap

class PlayerDao : BaseEntityDao<Player>("PlayerDao"), IPlayerDao {

    private var _playerOne: Player? = null

    override fun create(collidableId: Int, invincibleTimerId: Int, weaponId: Int, score: Int, health: Byte): Int {
        val id = nextId
        val player = Player(id, collidableId, invincibleTimerId, weaponId, score, health)

        _store.put(id, player)
        if (_playerOne == null) {
            _playerOne = player
        }

        return id
    }

    override fun retrieve(): Iterable<IPlayer> {
        return _store.values
    }

    override fun retrieve(id: Int): IPlayer {
        if (!_store.containsKey(id)) {
            throw RuntimeException(String.format("Key {0} not found", id))
        }

        return _store[id]
    }

    override fun playerOne(): IPlayer {
        return _playerOne
    }

    override fun update(id: Int, weaponId: Int, score: Int, health: Byte) {
        if (!_store.containsKey(id)) {
            throw RuntimeException(String.format("Key {0} not found", id))
        }

        val player = _store[id]

        player.weaponId = weaponId
        player.score = score
        player.health = health
    }

    override fun delete(id: Int) {
        if (!_store.containsKey(id)) {
            throw RuntimeException(String.format("Key {0} not found", id))
        }

        _store.remove(id)
        if (_playerOne!!.id == id) {
            _playerOne = null
        }
    }
}
